### ISSUE TITLE
CPU performance monitoring

### ISSUE THREAD

**Issue by ktsaou** [netdata/netdata#3232]
@tycho after reading this: http://www.brendangregg.com/blog/2017-05-09/cpu-utilization-is-wrong.html

I found that the following command might be useful for providing information about the memory performance of the system (i.e. "is memory slowing down the system?").

It seem that the overhead of this running all the time is ignorable. i couldn't even measure it.

What do you think? Would that be useful?

```sh
# perf stat -a -A -r 0 -d -d -d -x \; -- sleep 1
CPU0;1002.070440;;cpu-clock;1002071824;100.00;0.997;CPUs utilized
CPU1;1002.110830;;cpu-clock;1002111886;100.00;0.997;CPUs utilized
CPU2;1002.109677;;cpu-clock;1002109513;100.00;0.997;CPUs utilized
CPU3;1002.110049;;cpu-clock;1002109085;100.00;0.997;CPUs utilized
CPU4;1002.143965;;cpu-clock;1002143066;100.00;0.997;CPUs utilized
CPU5;1002.187700;;cpu-clock;1002186319;100.00;0.997;CPUs utilized
CPU6;1002.189881;;cpu-clock;1002189067;100.00;0.997;CPUs utilized
CPU7;1002.238780;;cpu-clock;1002237747;100.00;0.998;CPUs utilized
CPU0;377;;context-switches;1002284364;100.00;0.376;K/sec
CPU1;317;;context-switches;1002319706;100.00;0.316;K/sec
CPU2;255;;context-switches;1002319664;100.00;0.254;K/sec
CPU3;270;;context-switches;1002374995;100.00;0.269;K/sec
CPU4;263;;context-switches;1002130170;100.00;0.262;K/sec
CPU5;191;;context-switches;1002169133;100.00;0.191;K/sec
CPU6;506;;context-switches;1002170507;100.00;0.505;K/sec
CPU7;408;;context-switches;1002221288;100.00;0.407;K/sec
CPU0;11;;cpu-migrations;1002271686;100.00;0.011;K/sec
CPU1;7;;cpu-migrations;1002313245;100.00;0.007;K/sec
CPU2;14;;cpu-migrations;1002313546;100.00;0.014;K/sec
CPU3;23;;cpu-migrations;1002359284;100.00;0.023;K/sec
CPU4;12;;cpu-migrations;1002411271;100.00;0.012;K/sec
CPU5;14;;cpu-migrations;1002450811;100.00;0.014;K/sec
CPU6;18;;cpu-migrations;1002453064;100.00;0.018;K/sec
CPU7;14;;cpu-migrations;1002502625;100.00;0.014;K/sec
CPU0;132;;page-faults;1002537637;100.00;0.132;K/sec
CPU1;204;;page-faults;1002555201;100.00;0.204;K/sec
CPU2;12;;page-faults;1002554434;100.00;0.012;K/sec
CPU3;266;;page-faults;1002554849;100.00;0.265;K/sec
CPU4;2984;;page-faults;1002556467;100.00;0.003;M/sec
CPU5;121;;page-faults;1002557828;100.00;0.121;K/sec
CPU6;199;;page-faults;1002557868;100.00;0.199;K/sec
CPU7;1362;;page-faults;1002558324;100.00;0.001;M/sec
CPU0;73751219;;cycles;386999104;38.60;0.074;GHz
CPU1;100597302;;cycles;387024985;38.61;0.100;GHz
CPU2;65415513;;cycles;387350110;38.64;0.065;GHz
CPU3;60549816;;cycles;387221978;38.63;0.060;GHz
CPU4;101152463;;cycles;387274585;38.63;0.101;GHz
CPU5;59938699;;cycles;387027984;38.61;0.060;GHz
CPU6;84478855;;cycles;387303306;38.63;0.084;GHz
CPU7;88413735;;cycles;387256578;38.63;0.088;GHz
CPU0;25711909;;instructions;464094628;46.29;0.35;insn per cycle
CPU1;61619869;;instructions;464053963;46.29;0.84;insn per cycle
CPU2;28437028;;instructions;464258779;46.31;0.39;insn per cycle
CPU3;28135947;;instructions;464233037;46.30;0.38;insn per cycle
CPU4;60148198;;instructions;464215687;46.30;0.82;insn per cycle
CPU5;32433283;;instructions;464194820;46.29;0.44;insn per cycle
CPU6;61066240;;instructions;464379177;46.31;0.83;insn per cycle
CPU7;59255626;;instructions;464384511;46.31;0.80;insn per cycle
CPU0;4902810;;branches;463615267;46.23;4.893;M/sec
CPU1;14534639;;branches;463700279;46.24;14.505;M/sec
CPU2;6344808;;branches;463763269;46.25;6.332;M/sec
CPU3;6142647;;branches;463659615;46.23;6.130;M/sec
CPU4;13321461;;branches;463745119;46.24;13.294;M/sec
CPU5;7243259;;branches;463842295;46.25;7.228;M/sec
CPU6;12449547;;branches;463895408;46.25;12.424;M/sec
CPU7;11816629;;branches;463777000;46.24;11.792;M/sec
CPU0;277699;;branch-misses;463766769;46.24;5.66;of all branches
CPU1;533305;;branch-misses;463646073;46.22;10.88;of all branches
CPU2;258123;;branch-misses;463800717;46.24;5.26;of all branches
CPU3;241096;;branch-misses;463923689;46.25;4.92;of all branches
CPU4;261750;;branch-misses;463832927;46.24;5.34;of all branches
CPU5;221738;;branch-misses;463734198;46.23;4.52;of all branches
CPU6;305025;;branch-misses;463853635;46.24;6.22;of all branches
CPU7;477233;;branch-misses;464011320;46.26;9.73;of all branches
CPU0;6172895;;L1-dcache-loads;464101364;46.27;6.160;M/sec
CPU1;18381001;;L1-dcache-loads;463816641;46.24;18.343;M/sec
CPU2;5887029;;L1-dcache-loads;464019344;46.26;5.875;M/sec
CPU3;5331350;;L1-dcache-loads;463930082;46.25;5.320;M/sec
CPU4;17530997;;L1-dcache-loads;464019865;46.26;17.495;M/sec
CPU5;8815311;;L1-dcache-loads;463747325;46.24;8.797;M/sec
CPU6;14046593;;L1-dcache-loads;463923051;46.25;14.018;M/sec
CPU7;13507904;;L1-dcache-loads;463865255;46.25;13.480;M/sec
CPU0;936273;;L1-dcache-load-misses;463672989;46.23;15.17;of all L1-dcache hits
CPU1;1837467;;L1-dcache-load-misses;463952367;46.26;29.77;of all L1-dcache hits
CPU2;561317;;L1-dcache-load-misses;463671355;46.23;9.09;of all L1-dcache hits
CPU3;459604;;L1-dcache-load-misses;463750258;46.24;7.45;of all L1-dcache hits
CPU4;1549510;;L1-dcache-load-misses;463612741;46.23;25.10;of all L1-dcache hits
CPU5;779493;;L1-dcache-load-misses;463888869;46.25;12.63;of all L1-dcache hits
CPU6;928209;;L1-dcache-load-misses;463588908;46.23;15.04;of all L1-dcache hits
CPU7;1802009;;L1-dcache-load-misses;463705387;46.24;29.19;of all L1-dcache hits
CPU0;476743;;LLC-loads;307745023;30.69;0.476;M/sec
CPU1;609988;;LLC-loads;307790072;30.69;0.609;M/sec
CPU2;300342;;LLC-loads;307821762;30.70;0.300;M/sec
CPU3;232654;;LLC-loads;307733620;30.69;0.232;M/sec
CPU4;373090;;LLC-loads;307773974;30.69;0.372;M/sec
CPU5;345868;;LLC-loads;307857435;30.70;0.345;M/sec
CPU6;431264;;LLC-loads;307904437;30.71;0.430;M/sec
CPU7;385286;;LLC-loads;307861994;30.70;0.384;M/sec
CPU0;85300;;LLC-load-misses;308205204;30.74;17.89;of all LL-cache hits
CPU1;150409;;LLC-load-misses;308392876;30.76;31.55;of all LL-cache hits
CPU2;80288;;LLC-load-misses;308417282;30.76;16.84;of all LL-cache hits
CPU3;56689;;LLC-load-misses;308241355;30.74;11.89;of all LL-cache hits
CPU4;91863;;LLC-load-misses;308292144;30.75;19.27;of all LL-cache hits
CPU5;106638;;LLC-load-misses;308521162;30.77;22.37;of all LL-cache hits
CPU6;134118;;LLC-load-misses;308502008;30.77;28.13;of all LL-cache hits
CPU7;79837;;LLC-load-misses;308301590;30.75;16.75;of all LL-cache hits
CPU0;2372965;;L1-icache-load-misses;307903201;30.71;;;;
CPU1;2811421;;L1-icache-load-misses;308008840;30.72;;;;
CPU2;1407269;;L1-icache-load-misses;307728615;30.69;;;;
CPU3;1351176;;L1-icache-load-misses;307918305;30.71;;;;
CPU4;1775237;;L1-icache-load-misses;307957663;30.72;;;;
CPU5;1932934;;L1-icache-load-misses;308070738;30.73;;;;
CPU6;2574347;;L1-icache-load-misses;307814559;30.70;;;;
CPU7;2725306;;L1-icache-load-misses;307969966;30.72;;;;
CPU0;12105259;;dTLB-loads;308770690;30.80;12.080;M/sec
CPU1;12264999;;dTLB-loads;308552465;30.78;12.240;M/sec
CPU2;7955745;;dTLB-loads;308759344;30.80;7.939;M/sec
CPU3;6283968;;dTLB-loads;308647798;30.79;6.271;M/sec
CPU4;13540296;;dTLB-loads;308833048;30.81;13.512;M/sec
CPU5;8901754;;dTLB-loads;308627377;30.79;8.883;M/sec
CPU6;15059405;;dTLB-loads;308794573;30.80;15.028;M/sec
CPU7;12781753;;dTLB-loads;308692721;30.79;12.755;M/sec
CPU0;75356;;dTLB-load-misses;308929173;30.82;0.62;of all dTLB cache hits
CPU1;122211;;dTLB-load-misses;308356440;30.76;1.01;of all dTLB cache hits
CPU2;63427;;dTLB-load-misses;308301507;30.76;0.52;of all dTLB cache hits
CPU3;58137;;dTLB-load-misses;308431941;30.77;0.48;of all dTLB cache hits
CPU4;69793;;dTLB-load-misses;308417467;30.77;0.58;of all dTLB cache hits
CPU5;79347;;dTLB-load-misses;308373782;30.77;0.66;of all dTLB cache hits
CPU6;84996;;dTLB-load-misses;308305218;30.77;0.70;of all dTLB cache hits
CPU7;98797;;dTLB-load-misses;308402952;30.78;0.82;of all dTLB cache hits
CPU0;83852;;iTLB-loads;307948786;30.73;0.084;M/sec
CPU1;87001;;iTLB-loads;308239790;30.76;0.087;M/sec
CPU2;12316;;iTLB-loads;307931865;30.73;0.012;M/sec
CPU3;29302;;iTLB-loads;308060560;30.75;0.029;M/sec
CPU4;7051;;iTLB-loads;307946940;30.74;0.007;M/sec
CPU5;63442;;iTLB-loads;308170689;30.76;0.063;M/sec
CPU6;24204;;iTLB-loads;307924523;30.74;0.024;M/sec
CPU7;26075;;iTLB-loads;308072562;30.75;0.026;M/sec
CPU0;33994;;iTLB-load-misses;308554630;30.80;40.54;of all iTLB cache hits
CPU1;72460;;iTLB-load-misses;308432289;30.79;86.41;of all iTLB cache hits
CPU2;26008;;iTLB-load-misses;308678781;30.81;31.02;of all iTLB cache hits
CPU3;20055;;iTLB-load-misses;308646491;30.81;23.92;of all iTLB cache hits
CPU4;22635;;iTLB-load-misses;308557209;30.80;26.99;of all iTLB cache hits
CPU5;27879;;iTLB-load-misses;308430542;30.79;33.25;of all iTLB cache hits
CPU6;28949;;iTLB-load-misses;308685917;30.82;34.52;of all iTLB cache hits
CPU7;43158;;iTLB-load-misses;308645406;30.81;51.47;of all iTLB cache hits
```

**Comment by ktsaou** [comment]
This is probably a better command:

```sh
perf stat -a -A -r 0 -d -d -d -x \; -I 1000
```

it does not spawn any `sleep` commands.

**Comment by ktsaou** [comment]
hm... it needs 1% CPU of a single core on my laptop, but it can be controlled with `-I ms` to lower the data collection frequency...

**Comment by tycho** [comment]
It think it's probably a useful approach. It's very nice that it gives all the data we'd need to create a graph and in such a nice simple semicolon-delimited format, even including the units like `M/sec`, `GHz`, etc. 

I think a lot of the data it gathers is redundant because of other sources we already look at (e.g. context switches, although the current context switch graph in netdata is the aggregate of that stat for all CPUs I think).

I'm not sure how a long-running `perf stat` will interact with other consumers of the perf events subsystem, but it may be fine. I'd be concerned if a user runs `perf record` to profile something and gets bogus results just because netdata was running a `perf stat` the entire time in the background. But that may not be a problem in practice, I'm not really sure.

I think that blog post is kind of misleading, though. `CPU%` is a higher level metric than just "time the CPU hardware was busy". It's more like "time the kernel allocated to tasks". Nobody really expects `CPU%` to express any kind of efficiency metric for the tasks that are running.

**Comment by tycho** [comment]
By the way, might want to add `-T` to the command line to get TSX metrics (if present) as well.

**Comment by ktsaou** [comment]
What that page says, is that when `instructions per cycle` is below 1 the applications running are memory bound, while when it is above 1 the applications are CPU bound. In both cases the system reports 100% cpu utilization.

If this is true, it is very interesting.

**Comment by tycho** [comment]
Could also just be poorly optimized code. You can be CPU bound but still
have a crappy IPC (e.g. poor instruction mix/ordering, mispredicted
branches and other pipeline stalls, etc).

It *is* true that a very low IPC *might* suggest being memory-bound. I
think the TLB/cache statistics are more indicative of that (e.g. lots of
cache misses means lots of touching memory).

**Comment by ktsaou** [comment]
Can you help me understand it?


```sh
# perf stat -a -d -d -d -I 1000

   123.100852112        8008.478891      cpu-clock (msec)                                            
   123.100852112              4,702      context-switches          #    0.587 K/sec                  
   123.100852112                174      cpu-migrations            #    0.022 K/sec                  
   123.100852112              2,012      page-faults               #    0.251 K/sec                  
   123.100852112      5,109,360,888      cycles                    #    0.638 GHz                      (38.46%)
   123.100852112      5,574,865,332      instructions              #    0.53  insn per cycle           (46.15%)
   123.100852112      1,022,617,146      branches                  #  127.724 M/sec                    (46.15%)
   123.100852112          9,171,905      branch-misses             #    0.49% of all branches          (46.15%)
   123.100852112      1,674,419,284      L1-dcache-loads           #  209.134 M/sec                    (46.16%)
   123.100852112         65,228,240      L1-dcache-load-misses     #    2.25% of all L1-dcache hits    (46.16%)
   123.100852112         14,361,199      LLC-loads                 #    1.794 M/sec                    (30.77%)
   123.100852112          5,119,986      LLC-load-misses           #   15.58% of all LL-cache hits     (30.77%)
   123.100852112    <not supported>      L1-icache-loads                                             
   123.100852112         39,060,935      L1-icache-load-misses                                         (30.77%)
   123.100852112      1,715,804,346      dTLB-loads                #  214.303 M/sec                    (30.77%)
   123.100852112          3,198,231      dTLB-load-misses          #    0.11% of all dTLB cache hits   (30.77%)
   123.100852112          1,239,940      iTLB-loads                #    0.155 M/sec                    (30.77%)
   123.100852112            549,027      iTLB-load-misses          #   18.57% of all iTLB cache hits   (30.77%)
   123.100852112    <not supported>      L1-dcache-prefetches                                        
   123.100852112    <not supported>      L1-dcache-prefetch-misses                                   

   235.189485813        8004.014071      cpu-clock (msec)                                            
   235.189485813              5,624      context-switches          #    0.702 K/sec                  
   235.189485813                417      cpu-migrations            #    0.052 K/sec                  
   235.189485813              3,303      page-faults               #    0.413 K/sec                  
   235.189485813     11,625,372,580      cycles                    #    1.452 GHz                      (38.46%)
   235.189485813     13,941,138,788      instructions              #    1.34  insn per cycle           (46.15%)
   235.189485813      2,557,966,993      branches                  #  319.491 M/sec                    (46.15%)
   235.189485813         23,073,385      branch-misses             #    1.22% of all branches          (46.15%)
   235.189485813      4,111,076,600      L1-dcache-loads           #  513.475 M/sec                    (46.16%)
   235.189485813        139,839,799      L1-dcache-load-misses     #    4.72% of all L1-dcache hits    (46.15%)
   235.189485813         31,943,004      LLC-loads                 #    3.990 M/sec                    (30.77%)
   235.189485813         12,681,285      LLC-load-misses           #   38.79% of all LL-cache hits     (30.77%)
   235.189485813    <not supported>      L1-icache-loads                                             
   235.189485813         72,804,608      L1-icache-load-misses                                         (30.77%)
   235.189485813      4,229,901,867      dTLB-loads                #  528.317 M/sec                    (30.77%)
   235.189485813          6,912,391      dTLB-load-misses          #    0.23% of all dTLB cache hits   (30.77%)
   235.189485813          2,905,386      iTLB-loads                #    0.363 M/sec                    (30.77%)
   235.189485813            823,056      iTLB-load-misses          #   27.66% of all iTLB cache hits   (30.77%)
   235.189485813    <not supported>      L1-dcache-prefetches                                        
   235.189485813    <not supported>      L1-dcache-prefetch-misses                                   

```

The first block has instructions per cycle 0.53 while the second has 1.34.
How should I interpret TLB or cache misses?

**Comment by ktsaou** [comment]
hm... I guess: higher cache hits = instructions bound, lower cache hit = memory bound.

**Comment by tycho** [comment]
Yeah, exactly. Higher miss rate means it has to load from some higher level of memory. L1 miss means it had to load from L2 or higher, and LLC (Last-Level Cache) misses mean it's loading from RAM. If it's missing LLC all the time, then you're probably memory-bound.

Here's my output of `perf stat -a -d -d -d -I 3000` before and during a run of an OpenMP-threaded STREAM benchmark:
```
     6.002702533       24012.005270      cpu-clock (msec)
     6.002702533              9,003      context-switches          #    0.375 K/sec
     6.002702533                804      cpu-migrations            #    0.033 K/sec
     6.002702533             12,707      page-faults               #    0.529 K/sec
     6.002702533      2,213,175,789      cycles                    #    0.092 GHz                      (30.72%)
     6.002702533        903,956,672      instructions              #    0.43  insn per cycle           (38.40%)
     6.002702533        183,041,114      branches                  #    7.624 M/sec                    (38.47%)
     6.002702533          6,618,708      branch-misses             #    4.32% of all branches          (38.48%)
     6.002702533        258,891,888      L1-dcache-loads           #   10.783 M/sec                    (38.46%)
     6.002702533         22,269,380      L1-dcache-load-misses     #    9.64% of all L1-dcache hits    (38.49%)
     6.002702533          7,051,992      LLC-loads                 #    0.294 M/sec                    (30.76%)
     6.002702533          1,042,437      LLC-load-misses           #   14.59% of all LL-cache hits     (30.76%)
     6.002702533    <not supported>      L1-icache-loads
     6.002702533         28,598,685      L1-icache-load-misses                                         (30.80%)
     6.002702533        188,953,872      dTLB-loads                #    7.870 M/sec                    (30.79%)
     6.002702533          1,499,204      dTLB-load-misses          #    0.70% of all dTLB cache hits   (30.77%)
     6.002702533            149,391      iTLB-loads                #    0.006 M/sec                    (30.76%)
     6.002702533            358,571      iTLB-load-misses          #  241.33% of all iTLB cache hits   (30.75%)
     6.002702533    <not supported>      L1-dcache-prefetches
     6.002702533    <not supported>      L1-dcache-prefetch-misses
```
And here's during:
```
     9.003387957       24009.169834      cpu-clock (msec)
     9.003387957              8,412      context-switches          #    0.350 K/sec
     9.003387957                598      cpu-migrations            #    0.025 K/sec
     9.003387957          2,108,678      page-faults               #    0.088 M/sec
     9.003387957     52,130,258,903      cycles                    #    2.171 GHz                      (30.75%)
     9.003387957      9,924,347,013      instructions              #    0.53  insn per cycle           (38.44%)
     9.003387957      1,180,504,887      branches                  #   49.168 M/sec                    (38.48%)
     9.003387957          7,246,762      branch-misses             #    1.46% of all branches          (38.48%)
     9.003387957      2,741,036,393      L1-dcache-loads           #  114.165 M/sec                    (38.48%)
     9.003387957        648,417,834      L1-dcache-load-misses     #   60.73% of all L1-dcache hits    (38.49%)
     9.003387957         60,066,293      LLC-loads                 #    2.502 M/sec                    (30.78%)
     9.003387957         37,583,023      LLC-load-misses           #  151.63% of all LL-cache hits     (30.77%)
     9.003387957    <not supported>      L1-icache-loads
     9.003387957         24,228,439      L1-icache-load-misses                                         (30.78%)
     9.003387957      2,688,166,430      dTLB-loads                #  111.963 M/sec                    (30.77%)
     9.003387957          4,276,647      dTLB-load-misses          #    0.41% of all dTLB cache hits   (30.75%)
     9.003387957            300,675      iTLB-loads                #    0.013 M/sec                    (30.75%)
     9.003387957          4,478,357      iTLB-load-misses          # 2247.29% of all iTLB cache hits   (30.74%)
     9.003387957    <not supported>      L1-dcache-prefetches
     9.003387957    <not supported>      L1-dcache-prefetch-misses
```
Note how LLC misses go through the roof. (As an aside, I'm not actually sure what the `2247.29%` number for `iTLB-load-misses` is all about, or why any of these numbers can exceed 100%.)

Now compare that to a very computation-heavy (rather than memory-heavy) workload like an OpenMP-threaded n-body:
```
     6.003162590       24020.098896      cpu-clock (msec)
     6.003162590              6,533      context-switches          #    0.272 K/sec
     6.003162590                459      cpu-migrations            #    0.019 K/sec
     6.003162590             11,572      page-faults               #    0.482 K/sec
     6.003162590     87,958,200,993      cycles                    #    3.663 GHz                      (30.73%)
     6.003162590     99,563,798,628      instructions              #    1.13  insn per cycle           (38.43%)
     6.003162590      3,561,496,506      branches                  #  148.314 M/sec                    (38.44%)
     6.003162590          3,083,764      branch-misses             #    0.09% of all branches          (38.46%)
     6.003162590     20,661,933,893      L1-dcache-loads           #  860.438 M/sec                    (38.47%)
     6.003162590      4,690,196,385      L1-dcache-load-misses     #   22.73% of all L1-dcache hits    (38.48%)
     6.003162590         99,582,614      LLC-loads                 #    4.147 M/sec                    (30.79%)
     6.003162590          8,394,071      LLC-load-misses           #    8.90% of all LL-cache hits     (30.79%)
     6.003162590    <not supported>      L1-icache-loads
     6.003162590         23,684,789      L1-icache-load-misses                                         (30.79%)
     6.003162590     20,653,988,030      dTLB-loads                #  860.107 M/sec                    (30.79%)
     6.003162590         74,052,886      dTLB-load-misses          #    0.36% of all dTLB cache hits   (30.78%)
     6.003162590            370,419      iTLB-loads                #    0.015 M/sec                    (30.76%)
     6.003162590            351,780      iTLB-load-misses          #  103.83% of all iTLB cache hits   (30.75%)
     6.003162590    <not supported>      L1-dcache-prefetches
     6.003162590    <not supported>      L1-dcache-prefetch-misses
```
Very few LLC misses, but we are missing L1 data cache occasionally. The data set for n-body should be small enough to fit in L3 for certain, and probably even L2.

Oh, and the blog's "low IPC means you're memory bound" rule fits here. Just bear in mind that low IPC doesn't *exclusively* mean you're memory bound.

**Comment by ktsaou** [comment]
thanks!

> As an aside, I'm not actually sure what the 2247.29% number for iTLB-load-misses is all about, or why any of these numbers can exceed 100%

I tried to do the calculation myself, but it does not end up. It says `% of all LL-cache hits` so it should be `misses * 100 / hits`, which can go above 100%, but if you do the calculation:

```
  9.003387957         60,066,293      LLC-loads                 #    2.502 M/sec                    (30.78%)
     9.003387957         37,583,023      LLC-load-misses           #  151.63% of all LL-cache hits     (30.77%)
```

So:

```sh
# bc -l
loads=60066293
misses=37583023
hits=loads - misses
misses * 100 / hits
167.15995048762924610165
```

hm...

**Comment by ktsaou** [comment]
hm... something is wrong...

```
     9.003387957            300,675      iTLB-loads                #    0.013 M/sec                    (30.75%)
     9.003387957          4,478,357      iTLB-load-misses          # 2247.29% of all iTLB cache hits   (30.74%)
```

hits cannot be `loads - misses`.

**Comment by tycho** [comment]
Yeah exactly. I don't get the meaning there. Probably need to look up the
performance counter reference data from somewhere on Intel's website and
figure out how to interpret the values

**Comment by tycho** [comment]
Quoting from https://www.spinics.net/lists/linux-perf-users/msg04435.html

```
On Wed, 8 Nov 2017, Jan Pokorný wrote:

> Is iTLB-load-misses > 100% reported because of some deficiency
> of the platform, a bug in perf, or an expected behaviour?

If you look into the kernel, the two events being measured are
	[ C(RESULT_ACCESS) ] = 0x2085,  /* ITLB_MISSES.STLB_HIT */
	[ C(RESULT_MISS)   ] = 0xe85,   /* ITLB_MISSES.WALK_COMPLETED */

If you look this up in Intel Vol3b you can see that the "access" metric
measures First level ITLB misses that hit in the Second-level TLB

Wheras the miss metric is a list of accesses from any level that caused a 
walk of the page tables.

So the events don't necessarily sound like they match up very well, so 
it's beleivable you will get odd results when trying to calculate 
percentages based on them.

Vince
```
So the perf tool is conflating the two events.

Here are some other iTLB-related events we could be adding (`-e <event>`), that may be more relative to each other:
```
# perf list itlb

List of pre-defined events (to be used in -e):


frontend:
  frontend_retired.itlb_miss
       [Retired Instructions who experienced iTLB true miss (Precise event)]

virtual memory:
  itlb.itlb_flush
       [Flushing of the Instruction TLB (ITLB) pages, includes 4k/2M/4M pages]
  itlb_misses.miss_causes_a_walk
       [Misses at all ITLB levels that cause page walks]
  itlb_misses.stlb_hit
       [Instruction fetch requests that miss the ITLB and hit the STLB]
  itlb_misses.walk_active
       [Cycles when at least one PMH is busy with a page walk for code (instruction fetch) request. EPT page walk duration are excluded in Skylake]
  itlb_misses.walk_completed
       [Code miss in all TLB levels causes a page walk that completes. (All page sizes)]
  itlb_misses.walk_completed_1g
       [Code miss in all TLB levels causes a page walk that completes. (1G)]
  itlb_misses.walk_completed_2m_4m
       [Code miss in all TLB levels causes a page walk that completes. (2M/4M)]
  itlb_misses.walk_completed_4k
       [Code miss in all TLB levels causes a page walk that completes. (4K)]
  itlb_misses.walk_pending
       [Counts 1 per cycle for each PMH that is busy with a page walk for an instruction fetch request. EPT page walk duration are excluded in Skylake]
```
Someone in that same thread suggested trying `-e itlb_misses.stlb_hit -e frontend_retired.itlb_miss -e itlb_misses.walk_completed`

**Comment by ktsaou** [comment]
Just found this: https://github.com/opcm/pcm

```sh
# ./pcm.x 1

 Processor Counter Monitor  ($Format:%ci ID=%h$)


Number of physical cores: 4
Number of logical cores: 8
Number of online logical cores: 8
Threads (logical cores) per physical core: 2
Num sockets: 1
Physical cores per socket: 4
Core PMU (perfmon) version: 4
Number of core PMU generic (programmable) counters: 4
Width of generic (programmable) counters: 48 bits
Number of core PMU fixed counters: 3
Width of fixed counters: 48 bits
Nominal core frequency: 2800000000 Hz
Package thermal spec power: 45 Watt; Package minimum power: 0 Watt; Package maximum power: 0 Watt; 
Delay: 1
Trying to use Linux perf events...
Successfully programmed on-core PMU using Linux perf

Detected Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz "Intel(r) microarchitecture codename Kabylake" stepping 9

 EXEC  : instructions per nominal CPU cycle
 IPC   : instructions per CPU cycle
 FREQ  : relation to nominal CPU frequency='unhalted clock ticks'/'invariant timer ticks' (includes Intel Turbo Boost)
 AFREQ : relation to nominal CPU frequency while in active state (not in power-saving C state)='unhalted clock ticks'/'invariant timer ticks while in C0-state'  (includes Intel Turbo Boost)
 L3MISS: L3 cache misses 
 L2MISS: L2 cache misses (including other core's L2 cache *hits*) 
 L3HIT : L3 cache hit ratio (0.00-1.00)
 L2HIT : L2 cache hit ratio (0.00-1.00)
 L3MPI : number of L3 cache misses per instruction
 L2MPI : number of L2 cache misses per instruction
 READ  : bytes read from main memory controller (in GBytes)
 WRITE : bytes written to main memory controller (in GBytes)
 IO    : bytes read/written due to IO requests to memory controller (in GBytes); this may be an over estimate due to same-cache-line partial requests
 TEMP  : Temperature reading in 1 degree Celsius relative to the TjMax temperature (thermal headroom): 0 corresponds to the max temperature
 energy: Energy in Joules


 Core (SKT) | EXEC | IPC  | FREQ  | AFREQ | L3MISS | L2MISS | L3HIT | L2HIT | L3MPI | L2MPI | TEMP

   0    0     0.01   0.62   0.02    0.54     100 K    212 K    0.49    0.49    0.00    0.01     51
   1    0     0.01   0.71   0.02    0.48     101 K    204 K    0.47    0.43    0.00    0.00     44
   2    0     0.02   0.75   0.03    0.50      80 K    215 K    0.61    0.47    0.00    0.00     51
   3    0     0.01   0.45   0.01    0.37      59 K    123 K    0.48    0.40    0.00    0.01     54
   4    0     0.01   0.73   0.01    0.44      38 K     92 K    0.57    0.61    0.00    0.00     51
   5    0     0.05   1.09   0.05    0.79     137 K    438 K    0.68    0.70    0.00    0.00     44
   6    0     0.01   0.50   0.03    0.41      83 K    257 K    0.64    0.43    0.00    0.01     51
   7    0     0.02   0.63   0.03    0.63      76 K    185 K    0.57    0.51    0.00    0.00     54
---------------------------------------------------------------------------------------------------------------
 SKT    0     0.02   0.74   0.03    0.53     676 K   1728 K    0.59    0.56    0.00    0.00     44
---------------------------------------------------------------------------------------------------------------
 TOTAL  *     0.02   0.74   0.03    0.53     676 K   1728 K    0.59    0.56    0.00    0.00     N/A

 Instructions retired:  425 M ; Active cycles:  575 M ; Time (TSC): 2811 Mticks ; C0 (active,non-halted) core residency: 4.80 %

 C1 core residency: 9.78 %; C3 core residency: 0.20 %; C6 core residency: 6.56 %; C7 core residency: 78.65 %;
 C2 package residency: 27.24 %; C3 package residency: 38.50 %; C6 package residency: 0.00 %; C7 package residency: 0.00 %; C8 package residency: 0.00 %; C9 package residency: 0.00 %; C10 package residency: 0.00 %;

 PHYSICAL CORE IPC                 : 1.48 => corresponds to 37.03 % utilization for cores in active state
 Instructions per nominal CPU cycle: 0.04 => corresponds to 0.95 % core utilization over time interval
 SMI count: 56
---------------------------------------------------------------------------------------------------------------
MEM (GB)->|  READ |  WRITE |   IO   | CPU energy |
---------------------------------------------------------------------------------------------------------------
 SKT   0     2.14     0.33     1.38       5.76     
---------------------------------------------------------------------------------------------------------------

```

**Comment by anayrat** [comment]
How can we help you on this issue?

**Comment by cakrit** [comment]
@vlkobal this is a feature request to add new metrics for CPUs. It's very interesting and the article that Costas referenced in the beginning contains all the background. I haven't read the entire thread to the required level of detail to understand precisely what we will collect, how we'll organize the info in charts and what alarms we need, but please try to make a suggestion and @anayrat can help us specify it precisely so we can plan it.

**Comment by ilyam8** [comment]
@vlvkobal

**Comment by Ferroin** [comment]
The exact metrics are likely _extremely_ dependent on the CPU itself, as most of the stuff that `perf` scrapes is actual hardware performance counters.  It's also dependent on certain kernel options as well (though many distros build their kernels with those options).

Additionally, this should probably be done with it's own SUID helper program.  On sane systems, only root can access perf data.

**Comment by vlvkobal** [comment]
I've been searching in vain for grouping rules for the `perf_event_open()` system call. Does anybody know something about it on the spot?

**Comment by cakrit** [comment]
You mean what is says [here](http://man7.org/linux/man-pages/man2/perf_event_open.2.html)?

> The group_fd argument allows event groups to be created.  An event
       group has one event which is the group leader.  The leader is created
       first, with group_fd = -1.  The rest of the group members are created
       with subsequent perf_event_open() calls with group_fd being set to
       the file descriptor of the group leader.  (A single event on its own
       is created with group_fd = -1 and is considered to be a group with
       only 1 member.)  An event group is scheduled onto the CPU as a unit:
       it will be put onto the CPU only if all of the events in the group
       can be put onto the CPU.  This means that the values of the member
       events can be meaningfully compared—added, divided (to get ratios),
       and so on—with each other, since they have counted events for the
       same set of executed instructions.

**Comment by vlvkobal** [comment]
The problem I faced with is that I don't know for sure which events can be put into one group without conflicts. I figured out some working combinations experimentally, but it would be better to use a more reliable source of information.

**Comment by cakrit** [comment]
@anayrat and @tycho can you help out here?

**Comment by Saruspete** [comment]
Hello there,

In my (somewhat) old implementation of Intel' PCM (the tool @ktsaou proposed) into netdata, I did this: https://github.com/Saruspete/netdata.pcm (I've asked AMD to provide an equivalent to it).

That's awk+bash, so maybe not what you'll want, but for me it's working in my (critical) production since a few months. No fancy tweaks, nor playing with perf (where I had some fancy issues with on my rhel6/7 servers). Also lower overhead than perf.

I've started to modify directly pcm, but it's way harder than my quick script. But I would be able to select more precisely the counters I want (eg, don't care about cstates, so let's have per-dimm memory access).
